core/remotes/docker/fetcher.go

```diff
@@ -262,74 +437,122 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+	if parallelism > 1 && chunkSize > 512 && req.body == nil {
+		// If we have a content length, we can use multiple requests to fetch
+		// the content in parallel. This will make download of bigger bodies
+		// faster, at the cost of parallelism more requests and max
+		// ~(max_parallelism * goroutine footprint) memory usage. The goroutine
+		// footprint should be: the goroutine stack + pipe buffer size
+		numChunks := remaining / chunkSize
+		if numChunks*chunkSize < remaining {
+			numChunks++
 		}
-		var registryErr Errors
-		if err := json.NewDecoder(resp.Body).Decode(&registryErr); err != nil || registryErr.Len() < 1 {
-			return nil, fmt.Errorf("unexpected status code %v: %v", req.String(), resp.Status)
+		if numChunks < parallelism {
+			parallelism = numChunks
 		}
-		return nil, fmt.Errorf("unexpected status code %v: %s - Server message: %s", req.String(), resp.Status, registryErr.Error())
-	}
-	if offset > 0 {
-		cr := resp.Header.Get("content-range")
-		if cr != "" {
-			if !strings.HasPrefix(cr, fmt.Sprintf("bytes %d-", offset)) {
-				return nil, fmt.Errorf("unhandled content range in response: %v", cr)
-
-			}
-		} else {
-			// TODO: Should any cases where use of content range
-			// without the proper header be considered?
-			// 206 responses?
-
-			// Discard up to offset
-			// Could use buffer pool here but this case should be rare
-			n, err := io.Copy(io.Discard, io.LimitReader(resp.Body, offset))
-			if err != nil {
-				return nil, fmt.Errorf("failed to discard to offset: %w", err)
-			}
-			if n != offset {
-				return nil, errors.New("unable to discard to offset")
+		queue := make(chan int64, parallelism)
+		stopChan := make(chan struct{})
+		readers, writers := make([]io.Reader, numChunks), make([]*pipeWriter, numChunks)
+		bufPool := newbufferPool(chunkSize)
+		for i := range numChunks {
+			readers[i], writers[i] = newPipeWriter(bufPool)
+		}
+		go func() {
+			for i := range numChunks {
+				queue <- i
 			}
-
+			close(queue)
+		}()
+		r.Release(1)
+		for range parallelism {
+			go func() {
+				for i := range queue { // first in first out
+					copy := func() error {
+						if err := r.Acquire(ctx, 1); err != nil {
+							return err
+						}
+						defer r.Release(1)
+						select {
+						case <-stopChan:
+							return errors.New("another worker failed")
+						case <-ctx.Done():
+							return ctx.Err()
+						default:
+						}
+						if i == 0 {
+							_, err = io.Copy(writers[i], io.LimitReader(resp.Body, chunkSize))
+							_ = resp.Body.Close()
+							_ = writers[i].CloseWithError(err)
+							return nil
+						}
+						reqClone := req.clone()
+						reqClone.setOffset(offset + i*chunkSize)
+						nresp, err := reqClone.doWithRetries(ctx,
+							withErrorCheck,
+						)
+						if err != nil {
+							_ = writers[i].CloseWithError(err)
+							select {
+							case <-stopChan:
+								return errors.New("another worker failed")
+							default:
+								close(stopChan)
```

Comment:
This is still racy, multiple goroutines may hit this condition and cause a panic. There is no guarantee in one goroutines selects default and closes before another goroutine selects the default.

Reply:
ditto here, I put the close in a sync.Once to be sure, also made the stop close the context. LMK if anything as well.